In [1]:
import pandas as pd
from sklearn import model_selection
from sklearn import metrics
from sklearn.naive_bayes import GaussianNB

In [2]:
# The source data does not have any column names. Read the documentation about the data to identify the column names
# supplying list of names as input to create the dataframe

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/pima-indians-diabetes/pima-indians-diabetes.data"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
#dataframe = pandas.read_csv(url, names=names)
dataframe = pd.read_csv("D:\ML-Data\pima-indians-diabetes.data" , names=names)

In [4]:
dataframe.head(100)

preg  plas  pres  skin  test  mass   pedi  age  class
0      6   148    72    35     0  33.6  0.627   50      1
1      1    85    66    29     0  26.6  0.351   31      0
2      8   183    64     0     0  23.3  0.672   32      1
3      1    89    66    23    94  28.1  0.167   21      0
4      0   137    40    35   168  43.1  2.288   33      1
5      5   116    74     0     0  25.6  0.201   30      0
6      3    78    50    32    88  31.0  0.248   26      1
7     10   115     0     0     0  35.3  0.134   29      0
8      2   197    70    45   543  30.5  0.158   53      1
9      8   125    96     0     0   0.0  0.232   54      1
10     4   110    92     0     0  37.6  0.191   30      0
11    10   168    74     0     0  38.0  0.537   34      1
12    10   139    80     0     0  27.1  1.441   57      0
13     1   189    60    23   846  30.1  0.398   59      1
14     5   166    72    19   175  25.8  0.587   51      1
15     7   100     0     0     0  30.0  0.484   32      1
16     0   118    84    47   230  45.8  0.551   31      1
17     7   107    74     0     0  29.6  0.254   31      1
18     1   103    30    38    83  43.3  0.183   33      0
19     1   115    70    30    96  34.6  0.529   32      1
20     3   126    88    41   235  39.3  0.704   27      0
21     8    99    84     0     0  35.4  0.388   50      0
22     7   196    90     0     0  39.8  0.451   41      1
23     9   119    80    35     0  29.0  0.263   29      1
24    11   143    94    33   146  36.6  0.254   51      1
25    10   125    70    26   115  31.1  0.205   41      1
26     7   147    76     0     0  39.4  0.257   43      1
27     1    97    66    15   140  23.2  0.487   22      0
28    13   145    82    19   110  22.2  0.245   57      0
29     5   117    92     0     0  34.1  0.337   38      0
..   ...   ...   ...   ...   ...   ...    ...  ...    ...
70     2   100    66    20    90  32.9  0.867   28      1
71     5   139    64    35   140  28.6  0.411   26      0
72    13   126    90     0     0  43.4  0.583   42      1
73     4   129    86    20   270  35.1  0.231   23      0
74     1    79    75    30     0  32.0  0.396   22      0
75     1     0    48    20     0  24.7  0.140   22      0
76     7    62    78     0     0  32.6  0.391   41      0
77     5    95    72    33     0  37.7  0.370   27      0
78     0   131     0     0     0  43.2  0.270   26      1
79     2   112    66    22     0  25.0  0.307   24      0
80     3   113    44    13     0  22.4  0.140   22      0
81     2    74     0     0     0   0.0  0.102   22      0
82     7    83    78    26    71  29.3  0.767   36      0
83     0   101    65    28     0  24.6  0.237   22      0
84     5   137   108     0     0  48.8  0.227   37      1
85     2   110    74    29   125  32.4  0.698   27      0
86    13   106    72    54     0  36.6  0.178   45      0
87     2   100    68    25    71  38.5  0.324   26      0
88    15   136    70    32   110  37.1  0.153   43      1
89     1   107    68    19     0  26.5  0.165   24      0
90     1    80    55     0     0  19.1  0.258   21      0
91     4   123    80    15   176  32.0  0.443   34      0
92     7    81    78    40    48  46.7  0.261   42      0
93     4   134    72     0     0  23.8  0.277   60      1
94     2   142    82    18    64  24.7  0.761   21      0
95     6   144    72    27   228  33.9  0.255   40      0
96     2    92    62    28     0  31.6  0.130   24      0
97     1    71    48    18    76  20.4  0.323   22      0
98     6    93    50    30    64  28.7  0.356   23      0
99     1   122    90    51   220  49.7  0.325   31      1

[100 rows x 9 columns]

In [5]:
dataframe.describe()

preg        plas        pres        skin        test        mass  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   
75%      6.000000  140.250000   80.000000   32.000000  127.250000   36.600000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000   

             pedi         age       class  
count  768.000000  768.000000  768.000000  
mean     0.471876   33.240885    0.348958  
std      0.331329   11.760232    0.476951  
min      0.078000   21.000000    0.000000  
25%      0.243750   24.000000    0.000000  
50%      0.372500   29.000000    0.000000  
75%      0.626250   41.000000    1.000000  
max      2.420000   81.000000    1.000000

In [6]:
# Separate the independent attributes i.e. columns 0 to 8 and store them in X array
# Store the target column (last column) into Y array

array = dataframe.values
X = array[:,0:8] # select all rows and first 9 columns which are the attributes
Y = array[:,8]   # select all rows and the 10th column which is the classification "Yes", "No" for diabeties

In [7]:
# Create the training and test data set in the ratio of 70:30 respectively. Can be any other ratio...

test_size = 0.30 # taking 70:30 training and test set
seed = 7  # Random numbmer seeding for reapeatability of the code

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, Y, test_size=test_size, random_state=seed)

In [8]:
# Invoking the NB Gaussian function to create the model
# fitting the model in the training data set
model = GaussianNB()
model.fit(X_train, y_train)

GaussianNB(priors=None)

In [9]:
# make predictions
expected = y_test
predicted = model.predict(X_test)
# summarize the fit of the model
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

             precision    recall  f1-score   support

        0.0       0.80      0.79      0.79       147
        1.0       0.64      0.65      0.65        84

avg / total       0.74      0.74      0.74       231

[[116  31]
 [ 29  55]]


Precision: Within a given set of positively-labeled results, the fraction that were true positives = tp/(tp + fp)
Recall: Given a set of positively-labeled results, the fraction of all positives that were retrieved = tp/(tp + fn)
Accuracy: tp + tn / (tp + tn + fp +fn) But this measure can be dominated by larger class. Suppose 10, 90 and 80 of 90 is correctly predicted while only 2 of 0 is predicted correctly. Accuracy is 80+2 / 100 i.e. 82%

TO over come the dominance of the majority class, use weighted measure (not shown)

F is harmonic mean of precision and recal given by ((B^2 +1) PR) / (B^2P +R)
When B is set to 1 we get F1 = 2PR / (P+R)